### ISSUE TITLE
Want to create admin role for ApplicationDbInitializer unblock code.

### ISSUE THREAD

**Issue by arichika** [arichika/AspNet.Identity.Oracle#4]
@cengizani wrote...  
cengizani commented 2 hours ago  
Hi Arichika, Firstly, thank you so much for this project. It is good idea. I want to use your solution with our oracle db. I have completed register and login process. But when I want to user administration and group management then I have an ApplicationDbInitializer problem. So, when unblock code start with "public class ApplicationDbInitializer : " in IdentityConfig.cs and IdentityModels.cs " Database.SetInitializer<" code block then have an error 3 Message:"The type 'IdentitySample.Models.ApplicationDbContext' cannot be used as type parameter 'TContext' in the generic type or method 
'System.Data.Entity.DropCreateDatabaseIfModelChanges'. 
There is no implicit reference conversion from 'IdentitySample.Models.ApplicationDbContext' to 'System.Data.Entity.DbContext'." 
Would you like the help me for this problem? Thanks again.
  
@cengizani wrote...  
cengizani commented 44 minutes ago  
I found something, apologizing is made;
In IdentityModels.cs ApplicationDbContext class inherit from OracleDatabase. 
"public class ApplicationDbContext : OracleDatabase" 
in some examples on the Internet an inherit from IdentityDbContext with applicationuser
I wonder how it's done in the original MySql instance of info?
"public class ApplicationDbContext : IdentityDbContext"
or
"public class ApplicationDbContext : IdentityDbContext"

@arichika wrote...  
arichika commented 2 minutes ago  
"AspNet.Identity.Oracle" has been designed so that it does not use any of Entity Framework.
So, I think that "IdentityDbContext" can not be used. 
Be in this code there is a bug... 
https://github.com/arichika/AspNet.Identity.Oracle/blob/master/Src/SampleWebSite/App_Start/IdentityConfig.cs#L136 
"new RoleStore(param)" requires OracleDatabase object. 
https://github.com/arichika/AspNet.Identity.Oracle/blob/master/Src/AspNet.Identity.Oracle/RoleStore.cs#L46 
The code might need cast likes  
https://github.com/arichika/AspNet.Identity.Oracle/blob/master/Src/SampleWebSite/App_Start/IdentityConfig.cs#L29 
Want you to create admin role for ApplicationDbInitializer unblock code? Right? @cengizani
and I will create new issue ticket for this question.

**Comment by cengizani** [comment]
Yes sir, I want to create admin role. Unblock IdentityConfig.cs in public class ApplicationDbInitializer : DropCreateDatabaseIfModelChanges<ApplicationDbContext>
and also IdentityModels.cs in unlock code  
static ApplicationDbContext()
        {
            // Set the database intializer which is run once during application start
            // This seeds the database with admin user credentials and admin role
            Database.SetInitializer<ApplicationDbContext>(new ApplicationDbInitializer());
        }

**Comment by cengizani** [comment]
I checked my code again. Same as Your comments like. 
Clean solutions and rebuild code again but I have same error again.

IdenetityConfig.cs in:
public static ApplicationUserManager Create(IdentityFactoryOptions<ApplicationUserManager> options, IOwinContext context)
        {
            // var manager = new ApplicationUserManager(new UserStore<ApplicationUser>(context.Get<ApplicationDbContext>()));

            var manager = new ApplicationUserManager(new UserStore<ApplicationUser>(context.Get<ApplicationDbContext>() as OracleDatabase));

and 
Rolestore.cs in;
 /// <summary>
        /// Constructor that takes a Oracle Database as argument 
        /// </summary>
        /// <param name="database"></param>
        public RoleStore(OracleDatabase database)
        {
            Database = database;
            roleTable = new RoleTable(database);
        }

**Comment by cengizani** [comment]
![identity_oracle_applicationdbinitializer_dbcontext](https://cloud.githubusercontent.com/assets/8065972/9845536/5461a0ca-5ad4-11e5-9b0d-25988e77ba18.jpg)
![identity_oracle_applicationdbinitializer_dbcontext_identitymodels_cs](https://cloud.githubusercontent.com/assets/8065972/9845563/8f7d05aa-5ad4-11e5-8855-6200a8ef9013.jpg)
![identity_oracle_applicationdbinitializer_dbcontext_rolestore_cs](https://cloud.githubusercontent.com/assets/8065972/9845576/b6579a6e-5ad4-11e5-9695-6335b8b748c3.jpg)
![identity_oracle_applicationusermanager_identityconfig_cs](https://cloud.githubusercontent.com/assets/8065972/9845604/ea495010-5ad4-11e5-8f3e-a34d3e047cdc.jpg)

**Comment by arichika** [comment]
@cengizani  Because I have to restore the development environment, please wait for a while. ;-)

**Comment by cengizani** [comment]
Ok sir. I am sorry to rush.

**Comment by cengizani** [comment]
A current photo from Istanbul: Bosphorus Bridge
First Bridge of Turkey Made in Japan
![istanbul_bosphorus_bridge](https://cloud.githubusercontent.com/assets/8065972/9846475/092dbfc4-5adb-11e5-8d13-a7662f02141e.jpg)

**Comment by cengizani** [comment]
An also second brigde made from Japan name is Fatih Sultan Mehmet Bridge
![fatih-sultan-mehmet-bridge](https://cloud.githubusercontent.com/assets/8065972/9846546/84698e52-5adb-11e5-9f9f-34e213838c17.jpg)

**Comment by arichika** [comment]
@cengizani  Beautiful pictures!  
In Project "Marmaray", I know that the Japanese companies had cooperated in the construction of the undersea tunnel of the Bosphorus.

**Comment by cengizani** [comment]
Yes new project but 1 hundered years ago planned

**Comment by arichika** [comment]
I'm thinking some better way now.

AspNet.Identity.Oracle  does not depend on the Entity Framework is designed.
So, it can not be used to initialize using DropCreateDatabaseIfModelChanges<TContext>. 

For the same reason, before you deploy your application, manually you must create the tables in the Oracle Database.

First, It will already have tables on the DB, At the time of deployment.
In this case, To create the Admin Role at startup, it will called only once at somewhere in the application of the startup process where will execute only once.

**Comment by arichika** [comment]
In the current design, you must create the tables in the Oracle Database manually.
Most simple solution is to run the sql script to insert a administrator and a management role(='Admin') at deployment time.

I think that's not cool.

AspNet.Identity.Oracle does not depend on the Entity Framework, also,  SampleWebSite dones not depend on the Microsoft.AspNet.Identity.EntityFramework. For this reason, ASP.NET never calling "Database.SetInitializer\<TContext\>();"

After all, it should work if you write an initialization process that is performed only once correctly.I'm considering better place to write it.

**Comment by cengizani** [comment]
Yes, I created the tables in the database first. After you register and login operations can do it successfully.
OracleIdentity.sql  in there is no any insert command in SQL file. The insert commands in the tables in the project with the problem solved?
Thank you very much.

I'm sorry for the catastrophe that lights up the mountain in Japan. I hope I have not been a problem. I'm working on a company is Digiturk. Digital satellite broadcasting, a company that makes.
In addition, Turkey and 4 different country's football and basketball broadcasts and of course series. United Kingdom in the Sky, a company like Dish or Direct Tv in America.

I think about you a lot today. Thank you so much for your contribution to this very valuable. If a request from Istanbul to try to help as I can. Again, thank you very much.

**Comment by arichika** [comment]
2 changes in develop branch.

https://github.com/arichika/AspNet.Identity.Oracle/blob/develop/Src/SampleWebSite/App_Start/Startup.Auth.cs#L23
https://github.com/arichika/AspNet.Identity.Oracle/blob/develop/Src/SampleWebSite/App_Start/IdentityConfig.cs#L160-L260

- Startup.Auth.cs#L16  
Add this. 
```
app.CreatePerOwinContext<ApplicationDbInitializer>(ApplicationDbInitializer.Create);
```

- IdentityConfig.cs#L160-L260  
For initialize db. now empty.
```
private void InitializeDb(IOwinContext context)
{
~
}
```

For initialize user, r.g. first administrator.
```
private static void InitializeIdentity(IOwinContext context)
{
~
}
```

ASP.NET Identity / Owin will calling ApplicationDbInitializer.Create sometimes. 
So, The ApplicationDbInitializer as a singleton, it is a global once processing.

https://github.com/arichika/AspNet.Identity.Oracle/blob/develop/Src/SampleWebSite/App_Start/IdentityConfig.cs#L215-L217

```
            const string name = "admin@admin.com";
            const string password = "Admin@123456";
            const string roleName = "Admin";
```

Create Role Admin if it does not exist, and Add user admin@admin.com (pw Admin@123456) if it does not exist, and Add to Role Admin if not already added.

**Comment by arichika** [comment]
@cengizani  I did merge to master branch and release. 
https://github.com/arichika/AspNet.Identity.Oracle/releases/tag/v1.0.1

Thank you for your thoughtfulness.
Turkey is a fascinating country.Someday, I want to go to Turkey.

**Comment by cengizani** [comment]
The final version of the project ran successfully with. I had a meeting. So I'm sorry I couldn't turn back now. But I think it's perfect. The final version of the project ran successfully with.


I want to thank you. So soon that you are interested in. In addition, attentive and quality work, thank you so much for making everything all over again.

![success](https://cloud.githubusercontent.com/assets/8065972/9874872/7ea3294c-5bb2-11e5-92a7-59cdcb55ea9e.jpg)

**Comment by arichika** [comment]
@cengizani  It was good. :+1:  Because there was your indicated, I was able to better. Thank you!
